### PR TITLE
KAFKA-3038; Future'based pseudo-async controller

### PR DESCRIPTION
**Pull request by resetius**

### CODE REVIEW

**Comment by resetius** [comment]
This PR based on https://github.com/apache/kafka/pull/750 but it uses Futures so code is match more simple.
Tested on high load cluster with ~100K partitions. The speedup of leader election is up to 100x.

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala [hunk 1]
```diff
@@ -102,6 +105,18 @@ class PartitionStateMachine(controller: KafkaController) extends Logging {
     info("Stopped partition state machine")
   }
 
+  def executeFutureCallbacks(listOfFutures: Iterable[Future[() => Unit]]): Unit = {
+    import scala.concurrent.duration._
+    val futureOfList = Future.sequence(listOfFutures)
```
Comment (ataraxer) [inline core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: [Реализация](https://github.yandex-team.ru/gist/ataraxer/3bbf9fd17815b57adeceb3f60bfa526d).
Тут.

File: core/src/main/scala/kafka/controller/ReplicaStateMachine.scala [hunk 2]
```diff
@@ -111,7 +113,21 @@ class ReplicaStateMachine(controller: KafkaController) extends Logging {
       info("Invoking state change to %s for replicas %s".format(targetState, replicas.mkString(",")))
       try {
         brokerRequestBatch.newBatch()
-        replicas.foreach(r => handleStateChange(r, targetState, callbacks))
+        // replicas.foreach(r => handleStateChange(r, targetState, callbacks))
+
+        val listOfFutures = replicas.map(r => handleStateChangeAsync(r, targetState, callbacks))
+
+        import scala.concurrent.duration._
+        val futureOfList = Future.sequence(listOfFutures)
```
Comment (ataraxer) [inline core/src/main/scala/kafka/controller/ReplicaStateMachine.scala]: И тут.

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala (hunk 1 above)
Comment (ataraxer) [inline core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: `Futures.sequence` [folds](https://github.com/scala/scala/blob/2.11.x/src/library/scala/concurrent/Future.scala#L499) over a provided list of futures chaining them via `flatMap` which doesn't allow GC to collect them individually after completion. Here's a simple implementation that doesn't has this problem: [gist](https://gist.github.com/ataraxer/7d49e49e90ba025dbcdf679c39297741).

File: core/src/main/scala/kafka/controller/ReplicaStateMachine.scala (hunk 2 above)
Comment (ataraxer) [inline core/src/main/scala/kafka/controller/ReplicaStateMachine.scala]: Same here.

**Review by ataraxer — CHANGES_REQUESTED** [review]
Sorry, hadn't noticed that this isn't internal GitHub Enterprise when following a link, please disregard previous comments.

File: core/src/main/scala/kafka/controller/ControllerChannelManager.scala [hunk 3]
```diff
@@ -280,7 +280,11 @@ class ControllerBrokerRequestBatch(controller: KafkaController) extends  Logging
 
     brokerIds.filter(_ >= 0).foreach { brokerId =>
       val result = leaderAndIsrRequestMap.getOrElseUpdate(brokerId, mutable.Map.empty)
-      result.put(topicPartition, PartitionStateInfo(leaderIsrAndControllerEpoch, replicas.toSet))
```
Comment (resetius) [inline core/src/main/scala/kafka/controller/ControllerChannelManager.scala:287]: Futures answers can be reordered so we must check leaderEpoch here

**Comment by resetius** [comment]
Rebased against trunk
ataraxer's suggestions applied

**Comment by resetius** [comment]
parallel paritition creation added

**Pull request closed without merging**